### PR TITLE
WIP/PoC Use python type hints for precondition

### PR DESCRIPTION
**Pull request by takac**

### CODE REVIEW

File: dpcontracts.py [hunk 1]
```diff
@@ -468,14 +469,14 @@ def arg_count(func):
...
-def condition(description, predicate, precondition=False, postcondition=False, instance=False):
+def condition(description, predicate, precondition=False, postcondition=False, instance=False, dynamic=False):
     assert isinstance(description, str), "contract descriptions must be strings"
     assert len(description) > 0, "contracts must have nonempty descriptions"
     assert isfunction(predicate), "contract predicates must be functions"
     assert not iscoroutinefunction(predicate), "contract predicates cannot be coroutines"
     assert precondition or postcondition, "contracts must be at least one of pre- or post-conditional"
-    assert arg_count(predicate) == (1 if precondition or instance else 2), \
-           "contract predicates must take the correct number of arguments"
+    # assert arg_count(predicate) == (1 if precondition or instance else 2), \
```
Comment (deadpixi) [inline dpcontracts.py]: Would it be possible to retain this assertion in a different way?

File: dpcontracts.py [hunk 2]
```diff
@@ -546,6 +552,124 @@ def inner(*args, **kwargs):
         return inner
     return func
 
+def pytypes():
```
Comment (deadpixi) [inline dpcontracts.py]: If possible, could we name this just "types"? That way the user could do `import types from contracts` or just use `contracts.types`, which (IMHO) seems cleaner.

File: dpcontracts.py [hunk 3]
```diff
@@ -546,6 +552,124 @@ def inner(*args, **kwargs):
...
+    """
```
Comment (deadpixi) [inline dpcontracts.py]: While having this here is probably the more correct way, for the sake of consistency could we move it into the module's docstring?

**Comment by deadpixi** [comment]
Looks great! If you could, please see the comments I left on the file.

**Review by takac — COMMENTED** [review]
Thanks for the feedback!

I'm cautious that there are non-obvious issues when using the `typing` library and would like to add a bit more depth to the exception that gets raised. Perhaps that's a follow up PR.

**Pull request closed without merging**